### PR TITLE
add env to spawned processes

### LINKED ISSUE #3876: [turbopack] Node process executed without process environment variables set
### What version of Turbopack are you using?

turbopack-230213.2

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Mac

### Describe the Bug

`process.env` has no shell variables when `next.config.js` is loaded inside `next-dev`.
I'm using turbopack from Next.js.

### Expected Behavior

`process.env` should have variables given from the shell.

### To Reproduce

1. Clone the repository.
2. `pnpm i && pnpm dev:turbo`
3. Open `http://localhost:3000/client-side` and you'll see the hydration error.
4. Set `kNextConfigCheckIfEnvIsDefined` to `true` in `next.config.js` and restart the server.
5. Open `http://localhost:3000` and you'll see the server crash.

These errors do not occur in the normal `next dev`.

I made a workaround for this but I'm not sure if this is safe (as `execute()` is called in many places).
https://github.com/3c1u/turbo/tree/fix-process-env-turbopack-node
With this patch, those errors will disappear.

### Reproduction Repo

https://github.com/3c1u/nextjs-env-injection

### PR DESCRIPTION
**Pull request by ForsakenHarmony**
Fixes WEB-517
Fixes #3876

### CODE REVIEW

**Comment by ForsakenHarmony** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4005** <a href="https://app.graphite.dev/github/pr/vercel/turbo/4005" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/vercel/turbo/4005?utm_source=stack-comment).

File: crates/next-core/js/src/entry/server-edge-api.tsx [hunk 1]
```diff
@@ -22,7 +22,7 @@ startHandler(async ({ request, response, query, params, path }) => {
       join(process.cwd(), ".next/server/pages", chunk)
     ),
     wasm: [],
-    env: [],
+    env: Object.keys(process.env),
```
Comment (ForsakenHarmony) [inline crates/next-core/js/src/entry/server-edge-api.tsx:25]: This was the actual fix I needed for `front`, but fixing the real env also isn't a bad thing

File: crates/next-core/src/env.rs [hunk 2]
```diff
@@ -47,6 +47,29 @@ pub async fn load_env(project_path: FileSystemPathVc) -> Result<ProcessEnvVc> {
     Ok(env)
 }
 
+#[turbo_tasks::function]
+pub async fn full_env(
```
Comment (jridgewell) [inline crates/next-core/src/env.rs:51]: Can you document how this differs from `env_for_js`

Comment (jridgewell) [reply crates/next-core/src/env.rs:51]: Also, where is this used?

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: crates/next-core/js/src/entry/server-edge-api.tsx (hunk 1 above)
Comment (sokra) [reply crates/next-core/js/src/entry/server-edge-api.tsx:25]: Technically it should only list env vars used in code, but I guess for dev mode it doesn't matter

File: crates/next-core/src/app_source.rs [hunk 3]
```diff
@@ -347,11 +347,11 @@ pub async fn create_app_source(
         output_path,
     );
 
+    let injected_env = env_for_js(EnvMapVc::empty().into(), false, next_config);
+    let env = CustomProcessEnvVc::new(env, next_config.env()).as_process_env();
```
Comment (sokra) [inline crates/next-core/src/app_source.rs:351]: Why is that needed? shouldn't next config env already be applied to `env`?

File: crates/next-core/src/env.rs (hunk 2 above)
Comment (sokra) [reply crates/next-core/src/env.rs:51]: `full_env` is a bit misleading as name. This is the `filtered_env` for client?

File: crates/next-core/src/page_source.rs [hunk 4]
```diff
@@ -217,11 +217,12 @@ pub async fn create_page_source(
     )
     .into();
 
+    let injected_env = env_for_js(EnvMapVc::empty().into(), false, next_config);
+    let env = CustomProcessEnvVc::new(env, next_config.env()).as_process_env();
```
Comment (sokra) [inline crates/next-core/src/page_source.rs:221]: Same

File: crates/turbopack-node/src/lib.rs [hunk 5]
```diff
@@ -215,7 +217,18 @@ pub async fn get_renderer_pool(
...
-    Ok(NodeJsPool::new(cwd, entrypoint, HashMap::new(), 4, debug).cell())
+    Ok(NodeJsPool::new(
+        cwd,
+        entrypoint,
+        env.read_all()
+            .await?
+            .iter()
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .collect(),
+        4,
```
Comment (sokra) [inline crates/turbopack-node/src/lib.rs:228]: This should probably not hard-code 4 child processes, but use `available_concurrency` instead

File: crates/next-core/js/src/entry/server-edge-api.tsx (hunk 1 above)
Comment (ForsakenHarmony) [reply crates/next-core/js/src/entry/server-edge-api.tsx:25]: that's true, maybe something we can address in the future

File: crates/next-core/src/env.rs (hunk 2 above)
Comment (ForsakenHarmony) [reply crates/next-core/src/env.rs:51]: This is unused actually, I was looking into different ways to implement this

I wanted to partially get rid of the injected env variables, but maybe I'll just undo that

File: crates/next-core/src/app_source.rs (hunk 3 above)
Comment (ForsakenHarmony) [reply crates/next-core/src/app_source.rs:351]: No, they're applied in `env_for_js` and I wanted the unescaped variables for the real env variables

Comment (jridgewell) [reply crates/next-core/src/app_source.rs:352]: I think we could reorder a few things, then?
```diff
-    let injected_env = env_for_js(EnvMapVc::empty().into(), false, next_config);
+    let env = CustomProcessEnvVc::new(env, next_config.env()).as_process_env();       
+    let injected_env = env_for_js(env, false, next_config);
```

And remove the `next_config.env()` part from `env_for_js`?

**Review by jridgewell — APPROVED** [review]

File: crates/turbopack-node/src/lib.rs (hunk 5 above)
Comment (ForsakenHarmony) [reply crates/turbopack-node/src/lib.rs]: done
